flang/lib/Semantics/resolve-directives.cpp

```diff
@@ -2368,7 +2368,8 @@ void OmpAttributeVisitor::CheckDataCopyingClause(
       // either 'private' or 'threadprivate' in enclosing context.
       if (!checkSymbol->test(Symbol::Flag::OmpThreadprivate) &&
           !(HasSymbolInEnclosingScope(symbol, currScope()) &&
-              symbol.test(Symbol::Flag::OmpPrivate))) {
+              (symbol.test(Symbol::Flag::OmpPrivate) ||
+                  symbol.test(Symbol::Flag::OmpFirstPrivate)))) {
```

Comment:
Two points here:
1. Does this require a change to the message here? It says PRIVATE or THREADPRIVATE. 
2. Can we add a test for this change in semantic check?

Reply:
1. As FIRSTPRIVATE provides a superset of the functionality provided by PRIVATE, I thought that using it would already imply that the variable is private. I also noticed that gfortran allows FIRSTPRIVATE but doesn't mention it explicitly:
> copyprivate variable 'aaa' is not threadprivate or private in outer context

 Do you think it would be better to mention FIRSTPRIVATE explicitly in the error message?

2. Sure, I'll add a semantic check for it.

Reply:
Yeah you are right: we don't need to explicitly call our firstprivate here.